runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractQueuedStreamView.java

```diff
@@ -105,20 +113,84 @@ protected ILogData getNextEntry(QueuedStreamContext context,
+                ILogData ld = read(thisRead);
+                if (getFrom == context.readQueue) {
+                    addToResolvedQueue(context, thisRead, ld);
+                }
+                return ld;
             }
-            return ld;
+        } else {
+            return getFromAddressMap(context, getFrom);
         }
 
         // None of the potential reads ended up being part of this
         // stream, so we return null.
         return null;
     }
 
+    private ILogData getFromAddressMap(QueuedStreamContext context, NavigableSet<Long> getFrom) {
+        boolean getNext;
+        do {
+            // If entries in queue to process
+            if (!getFrom.isEmpty()) {
+                final long thisRead = getFrom.pollFirst();
+                // Prepare a batch of addresses to perform a multi-read, if the current address is not available
+                // the batch of addresses will be read along this address, otherwise batch reads
+                // will be deferred to the next cache miss
+                List<Long> batchRead = getBatch(thisRead, getFrom);
+                log.trace("getNextEntry[{}]: reading {}, if not present read batch of addresses: {}",
+                        this, thisRead, batchRead);
+                ILogData ld = readRange(thisRead, batchRead);
+
+                if (getFrom == context.readQueue) {
+                    // Validate data entry belongs to this stream, if not skip
+                    // This verification protects against the case when tokens were issued for a stream by the sequencer,
+                    // but never written to, when a sequencer failover occurs and therefore these tokens are reassigned
+                    // to potentially a new stream, causing invalidation of theses addresses for this stream.
+                    if (ld.containsStream(this.id)) {
+                        addToResolvedQueue(context, thisRead, ld);
+                        return ld;
+                    } else {
+                        log.trace("getNextEntry[{}]: the data for address {} does not belong to this stream. Skip.",
+                                this, thisRead);
+                        getNext = true;
+                    }
+                } else {
+                    return ld;
+                }
+            } else {
+                // no more entries to retrieve or process in queue
+                getNext = false;
+            }
+        } while (getNext);
+
+        return null;
+    }
+
+    /**
+     * Prepare a batch of entries to be read, including the current address to retrieve.
+     *
+     * @param currentRead current address to retrieve.
+     * @param getFrom queue
+     * @return batch of entries.
+     */
+    private List<Long> getBatch(Long currentRead, NavigableSet<Long> getFrom) {
+        int counter = 1;
+        List<Long> batchRead = new ArrayList<>();
+        batchRead.add(currentRead);
+        Iterator<Long> it = getFrom.iterator();
```

Comment:
please use `getFrom.subSet(...)`

Reply:
For subSet I need to know the boundaries and we do not know the range we want.